persistent-evdev: init at unstable-2022-01-14

lodi opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Description of changes

Packaged aiberia/[persistent-evdev](https://github.com/aiberia/persistent-evdev).  This is a workaround that allows qemu/libvirtd evdev functionality to keep working even if a mouse or keyboard is hotplugged (or in my case, continue working in the presence of a flaky usb controller.)

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lodi committed:
persistent-evdev: init at unstable-2022-01-14

SuperSandro2000 commented on pkgs/servers/persistent-evdev/default.nix:
```diff
@@ -0,0 +1,39 @@
+    evdev pyudev
+  ];
+
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```

SuperSandro2000 commented on pkgs/servers/persistent-evdev/persistent-evdev-udev-rules.nix:
```diff
@@ -0,0 +1,26 @@
+  version = "unstable-2020-07-15";
+
+  udevRules = fetchurl {
+    url = "https://raw.githubusercontent.com/aiberia/persistent-evdev/master/udev/60-persistent-input-uinput.rules";
```
Please download a specific commit.

lodi commented:
Hi @SuperSandro2000 @aanderse, please let me know if this is ok to merge, thanks!

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,73 @@
+
+let
+  cfg = config.services.persistent-evdev;
+  configFile = pkgs.writeText "persistent-evdev-config.json" ''
```
maybe use `pkgs.formats.json` to generate this instead. the json format pretty-prints the resulting text automatically.

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,73 @@
+  cfg = config.services.persistent-evdev;
+  configFile = pkgs.writeText "persistent-evdev-config.json" ''
+    ${builtins.toJSON {
+      cache = "${cfg.cacheDir}";
```
```suggestion
      cache = cfg.cacheDir;
```

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,73 @@
+  options.services.persistent-evdev = {
+    enable = mkEnableOption "virtual input devices that persist even if the backing device is hotplugged";
+
+    cacheDir = mkOption {
```
prefer systemd `CacheDirectory` over manually handling these where possible. if the cache has to be moved out of `/var/cache/persistent-evdev` a systemd bind mount can be used.

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,73 @@
+      default = {};
+      type = types.attrsOf types.str;
+      description = ''
+        A set of virtual proxy device labels with backing physical device ids.
+
+        Physical devices should already exist in /dev/input/by-id/.  Proxy devices will be created
+        with a "uinput-" prefix.
+
+        See https://github.com/aiberia/persistent-evdev#example-usage-with-libvirt for example
+        usage of virtual devices with libvirt.  Remember to add uinput-* devices to the qemu
+        cgroup_device_acl list (nixos: virtualisation.libvirtd.qemu.verbatimConfig).
```
this is a docbook text, so links and paragraphs will have to use docbook formatting to show up in the manuals as you intended them to.

pennae commented on pkgs/servers/persistent-evdev/persistent-evdev-udev-rules.nix:
```diff
@@ -0,0 +1,26 @@
+
+  dontUnpack = true;
+
+  installPhase = ''
```
fetchurl supports both `postFetch` to run this shell fragment and `meta` to add metadata. no need to wrap it in another derivation only to add those two. (only snag is that fetchurl *doesn't* have a pname, only a `name`)

pennae reviewed (commented):
does this service have to run as `root/root`, or are more restricted credentials sufficient (eg `persistent-evdev/input`? would be nice to lock this down as far as possible.

lodi commented:
Thanks for the review @pennae.  I implemented the config file changes you recommended and fixed the doc strings.

* root: the script needs to write to /dev/uinput, which is root:root 600.  Let me know if there's a way to do this without root.
* fetchurl: I wasn't sure how to get this to work without putting the rules in their own package.  Can you sketch out a quick fragment of what you mean?

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,66 @@
+
+      serviceConfig = {
+        Restart = "on-failure";
+        ExecStart = "${pkgs.persistent-evdev}/bin/persistent-evdev.py %E/${"persistent-evdev.json"}";
```
```suggestion
        ExecStart = "${pkgs.persistent-evdev}/bin/persistent-evdev.py %E/persistent-evdev.json";
```
although it's generally preferrable to not dump configuration files into `/etc` whenever feasible. is there a utility that needs access to this file at runtime? if not it'd be better to remove the file from /etc (not least because with this unit changes in the config would not restart the service!)

pennae reviewed (commented):
have a look at eg https://github.com/NixOS/nixpkgs/blob/master/pkgs/data/fonts/open-fonts/default.nix for fetchurl usage. that does pretty much the same thing you want to do: download a file and move its contents to a specific location. you won't need the `recursiveHash` since you're dealing with a single file though.

lodi commented:
Ok I simplified things and got rid of the unnecessary udev package altogether.  Also the author [updated](https://github.com/aiberia/persistent-evdev/issues/1) the license to MIT.

pennae commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,65 @@
+        Restart = "on-failure";
+        ExecStart = "${pkgs.persistent-evdev}/bin/persistent-evdev.py ${configFile}";
+        CacheDirectory = "persistent-evdev";
+        User = "root";
```
```suggestion
```
that's the default anyway

pennae reviewed (approved):
LGTM, but would very much like someone with python packaging knowledge to just have a quick look at this too. the source repo seems very simple in this regard so this is probably fine, but can't really tell.

lodi commented:
Okay thanks @pennae, I squashed all the changes and requested a review from a few maintainers of other python-based nixos modules.

aszlig commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,64 @@
+      type = types.attrsOf types.str;
+      description = ''
+        A set of virtual proxy device labels with backing physical device ids.
+        </para>
+        <para>
```
You don't need this because this is done in [documentation postprocessing](https://github.com/NixOS/nixpkgs/blob/4042f7b3752e1420b40eb6e337d187cea666c1e3/nixos/lib/make-options-doc/postprocess-option-descriptions.xsl#L22-L24) already.

aszlig commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,64 @@
+        <para>
+        Physical devices should already exist in <literal>/dev/input/by-id/</literal>.  Proxy
+        devices will be automatically given a <literal>uinput-</literal> prefix.
+        </para>
+        <para>
```
See above.

aszlig commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,64 @@
+        A set of virtual proxy device labels with backing physical device ids.
+        </para>
+        <para>
+        Physical devices should already exist in <literal>/dev/input/by-id/</literal>.  Proxy
```
```suggestion
        Physical devices should already exist in <filename class="devicefile">/dev/input/by-id/</filename>.  Proxy
```

aszlig commented on nixos/modules/services/misc/persistent-evdev.nix:
```diff
@@ -0,0 +1,64 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Please try to avoid this: https://nix.dev/anti-patterns/language#with-attrset-expression

lodi commented:
Thanks @aszlig, I made your recommended changes.

lodi commented:
Hi @SuperSandro2000, I think this is ready to be merged; could you take a look?  Thanks.

jian-lin commented:
@lodi 

> root: the script needs to write to /dev/uinput, which is root:root 600. Let me know if there's a way to do this without root.

I find this PR when I am making a similar one. You don't have to run as root to write `/dev/uinput`. You can check my [PR][1] for how to do that. You may also be interested in those hardening options.

[1]: https://github.com/NixOS/nixpkgs/pull/182756
